### PR TITLE
KAFKA-13457: SocketChannel in Acceptor#accept is not closed upon IOException

### PR DESCRIPTION
**Pull request by functioner**
A patch for [KAFKA-13457](https://issues.apache.org/jira/browse/KAFKA-13457)
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 1]
```diff
@@ -734,6 +734,10 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
         val endThrottleTimeMs = e.startThrottleTimeMs + e.throttleTimeMs
         throttledSockets += DelayedCloseSocket(socketChannel, endThrottleTimeMs)
         None
+      case e: IOException =>
+        info(s"Encounter IOException", e)
```
Comment (showuon) [inline core/src/main/scala/kafka/network/SocketServer.scala]: nit: `Encountered IOException, closing connection.`

**Review by showuon — COMMENTED** [review]
@functioner , thanks for finding this issue and fixing it. LGTM! Left a minor comment. Thanks.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 2]
```diff
@@ -734,6 +734,10 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
...
+        closeSocket(socketChannel)
```
Comment (dajac) [inline core/src/main/scala/kafka/network/SocketServer.scala]: We might have to use `close(endPoint.listenerName, socketChannel)` here. Otherwise, we would not decrement the number of connections in the `connectionQuotas`. Could you double check?

**Review by dajac — COMMENTED** [review]
@functioner Thanks for the patch.

It seems that the fix does not fully match the description in the JIRA. In the JIRA, you wrote the following: `However, line 717 or the socketChannel operations within the try block may potentially throw an IOException as well, which is not handled.`.  And you refers to this line:
```
val socketChannel = serverSocketChannel.accept()     // line 717
```

The try/catch block where the `IOException` has been added does not wrap that line so it does not catch it. I suppose that the description of the JIRA is incorrect in this case because we can't close the `socketChannel` if we don't have it.

However, it seems to be that such issue could be caused by `socketChannel.configureBlocking(false)` which could also thrown an `IOException`. Am I getting this right?

Do you think that we could add a unit test for this bug?

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 2 above)
Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: @dajac You're right. I've updated the JIRA issue descriptions. I have a new commit using `close(endPoint.listenerName, socketChannel)`.
For the test, since it involves injecting an IOException, I'm going to use Byteman or AspectJ. Or do you have any suggestion?

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 1 above)
Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: @showuon Thanks for the comment. I've updated the info message with the new commit

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 2 above)
Comment (dajac) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Hmm... Good question. I don't think that we already use those library in the code base so it would be great if we could do it with what we already use. I am not sure if it is possible.

Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: I think Byteman is widely used in many other communities, because it's a very elegant way to do fault injection testing. For example, ZooKeeper community has been discussing it in https://github.com/apache/zookeeper/pull/123. I think adding Byteman dependency won't take much time.

Comment (dajac) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Another way would be to extract `serverSocketChannel.accept()` into a method. Then we could override that method in a test and return a mocked socket which throws an exception.

Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: It seems doable. I'm looking for a unit test in `core/src/test/scala/unit/kafka/network/SocketServerTest.scala` based on which I can modify & write the test. Do you have any suggestion? It seems we also need to modify lots of code in `core/src/main/scala/kafka/network/SocketServer.scala` to enable this test.

Comment (dajac) [reply core/src/main/scala/kafka/network/SocketServer.scala]: `testMaxConnectionsPerIpOverrides` might be a good starting point as it test something in the same area. Why do you think that we need to modify a log of code in `SocketServer`?

If we extract `serverSocketChannel.accept()` into a method. Then, we can create a `SocketServer` and overwrite the method. In the method, we could return a mocked Socket which throws an exception when `configureBlocking` is called. Then, we could try to establish a connection. Finally, we can verify that the `close` methods were called.

```scala
val overrideServer = new SocketServer(...) {
  override def accept(): Socket = {

  }
}
```

Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: There are multiple classes defined in `core/src/main/scala/kafka/network/SocketServer.scala`. The `accept` method belongs to `Acceptor`, rather than `SocketServer`, so we can't override `accept` the way you show. We need to override `accept` in `Acceptor`, and then use this `Acceptor` when the `SocketServer` is initialized. Therefore, we need to first override `createAcceptor` in `SocketServer`, and then override `accept` in `Acceptor`. In addition, we need to change `createAcceptor` & `accept` from private to protected; we also need to change some fields from private to protected.

I've pushed a new commit implementing a test using this idea. I've confirmed that this test will pass only when the IOException is properly handled as I pointed out.

Comment (dajac) [reply core/src/main/scala/kafka/network/SocketServer.scala]: You're right. I overlooked this. In this case, I wonder if we could test the Acceptor only. Would it work? That would simplify all of this.

Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Since the existing test cases don't have scaffold for testing `Acceptor`, we need to prepare lots of parameters for `Acceptor` constructor
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L550-L558
We also need to prepare lots of parameters for creating a `Processor` for the `Acceptor`
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L687-L703
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L574-L578
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L790-L806
However, those logic are implemented in `SocketServer`
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L268-L283
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L424-L445
Therefore, I think testing Acceptor requires much more code. Instead, testing `SocketServer` is simpler.

**Comment by functioner** [comment]
Are there any other issues or concern with the current patch and test? If any, I will resolve ASAP. Will this PR be merged soon by any chance?

**Comment by dajac** [comment]
@functioner It seems that the code does not compile. I seems to be due to https://github.com/apache/kafka/commit/30a9085d505ac301acee18e4cdcc98d8060e0388 which added an parameter to the `Acceptor`'s constructor. Could you rebase or merge trunk into your branch?

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 3]
```diff
@@ -734,9 +730,21 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
         val endThrottleTimeMs = e.startThrottleTimeMs + e.throttleTimeMs
         throttledSockets += DelayedCloseSocket(socketChannel, endThrottleTimeMs)
         None
+      case e: IOException =>
+        info(s"Encountered IOException, closing connection", e)
```
Comment (dajac) [inline core/src/main/scala/kafka/network/SocketServer.scala]: Should we log this as an error? Moreover, I would rephrase a bit the log: `Encountered an error while configuring the connection, closing it.`

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala [hunk 4]
```diff
@@ -867,6 +868,39 @@ class SocketServerTest {
     }
   }
 
+  @Test
+  def testExceptionInAcceptor(): Unit = {
+    val overrideNum = server.config.maxConnectionsPerIp + 1
+    val overrideProps = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = 0)
+    overrideProps.put(KafkaConfig.MaxConnectionsPerIpOverridesProp, s"localhost:$overrideNum")
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: I guess that we don't need this in the context of this test, do we?

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala [hunk 5]
```diff
@@ -867,6 +868,39 @@ class SocketServerTest {
...
+          }
+        }
+      }
+    }
+
+    try {
+      overrideServer.startup()
+      val conn = connect(overrideServer)
+      conn.setSoTimeout(3000)
+      assertEquals(-1, conn.getInputStream.read())
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala:898]: Could we also verify that the connection quota is correct? We can use `connectionQuotas.get` for this with the ip of the relevant listener.

**Review by dajac — COMMENTED** [review]
Thanks for the update. Left a few comments.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala (hunk 4 above)
Comment (functioner) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: We need the `overrideProps` to create the `overrideServer`. From the other tests, I didn't figure out what could be a better configuration, so I just use the configuration in `testMaxConnectionsPerIpOverrides`. Or maybe we can try the configuration in `testConnectionRatePerIp`. Do you have any suggestion?

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 3 above)
Comment (functioner) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Right. I've pushed a commit.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala (hunk 5 above)
Comment (functioner) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala:898]: Right. I've pushed a commit. I check if the quota is 1 after the accept, and if the quota is 0 after the exception thrown.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala (hunk 4 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: Sorry, I was not clear here. Yes, we do need `overrideProps`. However, I think that we don't need to set `MaxConnectionsPerIpOverridesProp`, do we? If not, we could remove it and also remove `overrideNum`.

Comment (functioner) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: You're right. I've pushed a new commit

**Review by dajac — APPROVED** [review]
LGTM, thanks for the patch.

**Comment by functioner** [comment]
Thanks everybody here for the suggestions and merging the patch!
